core/store.go

```diff
@@ -316,6 +316,13 @@ func setExpiry(obj *Obj, expDurationMs int64) {
 	expires[obj] = uint64(utils.GetCurrentTime().UnixMilli()) + uint64(expDurationMs)
 }
 
+// setUnixTimeExpiry sets the expiry time for an object.
+// This method is not thread-safe. It should be called within a lock.
+func setUnixTimeExpiry(obj *Obj, exUnixTimeSec int64) {
+	// convert unix-time-seconds to unix-time-milliseconds
+	expires[obj] = uint64(exUnixTimeSec * 1000)
```

Comment:
Acquire a lock on the expires map when modifying it.

Reply:
Don't modify this method, define another method which calls this one while acquiring the respective lock.
Additionally, you can consider calling the delete API as a part of that new method (within the same critical section) to make the operation atomic.

The flow should be:
1. Acquire lock
2. Check if expiry in the past, delete if yes.
3. If not, set Expiry.
4. Release lock.

Reply:
@JyotinderSingh expire map is defined like this
```
expires   map[*Obj]uint64 // Does not need to be thread-safe as it is only accessed by a single thread.
```

in the codebase I can see that we do not take lock on it, while modifying it

Reply:
I guess you can avoid a lock here since it is only accessed by the eval loop (and the loop is single-threaded). We will anyway be moving away from the current architecture in some time.

Also you're right, we can depend on the db for clearing out the key, no need to eagerly delete.